### PR TITLE
MedicalNetPerceptualSimilarity: Add multi-channel

### LINKED ISSUE #7567: Perceptual Loss with mednet 3D multiple channels
**Describe the bug**
Perceptual loss with 3D volumes and the medicalnet network types currently works with one channel only, as MedNet3D takes single channel 3D volumes as input. However the code in MedicalNetPerceptualSimilarity suggests that it is intended to work with multiple channels.

**To Reproduce**
Use PerceptualLoss with spatial_dims=3, is_fake_3d=False,  network_type='medicalnet_resnet50_23datasets' and  3D volumes with multiple channels, e.g. with fMRI, phase-cycle bSSFP, DTI, ... as data.

**Expected behavior**
MedicalNetPerceptualSimilarity passes the input channelwise and averages (the latter is already implemented).

**Environment**
<details>
  <summary>MONAI environment</summary>
  
================================
Printing MONAI config...
================================
MONAI version: 1.3.0
Numpy version: 1.26.4
Pytorch version: 2.2.1+cu121
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: 865972f7a791bf7b42efbcd87c8402bd865b329e
MONAI __file__: /home/<username>/workspace/UNet-bSSFP/unet-venv/lib/python3.11/site-packages/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
ITK version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 5.2.0
scikit-image version: NOT INSTALLED or UNKNOWN VERSION.
scipy version: 1.12.0
Pillow version: 10.2.0
Tensorboard version: 2.16.2
gdown version: 4.7.3
TorchVision version: 0.17.1+cu121
tqdm version: 4.66.2
lmdb version: NOT INSTALLED or UNKNOWN VERSION.
psutil version: 5.9.8
pandas version: 2.0.3
einops version: NOT INSTALLED or UNKNOWN VERSION.
transformers version: NOT INSTALLED or UNKNOWN VERSION.
mlflow version: NOT INSTALLED or UNKNOWN VERSION.
pynrrd version: NOT INSTALLED or UNKNOWN VERSION.
clearml version: NOT INSTALLED or UNKNOWN VERSION.

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

================================
Printing system config...
================================
System: Linux
Linux version: Arch Linux
Platform: Linux-6.7.4-zen1-1-zen-x86_64-with-glibc2.39
Processor: 
Machine: x86_64
Python version: 3.11.7
Process name: python
Command: ['python', '-c', 'import monai; monai.config.print_debug_info()']
Open files: []
Num physical CPUs: 24
Num logical CPUs: 32
Num usable CPUs: 32
CPU usage (%): [93.7, 62.3, 79.0, 40.4, 61.9, 78.7, 56.0, 98.4, 96.5, 19.7, 23.0, 97.6, 96.4, 99.2, 99.6, 30.3, 99.6, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0]
CPU freq. (MHz): 3461
Load avg. in last 1, 5, 15 mins (%): [87.6, 86.7, 82.9]
Disk usage (%): 44.8
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 62.6
Available memory (GB): 25.2
Used memory (GB): 32.0

================================
Printing GPU config...
================================
Num GPUs: 1
Has CUDA: True
CUDA version: 12.1
cuDNN enabled: True
NVIDIA_TF32_OVERRIDE: None
TORCH_ALLOW_TF32_CUBLAS_OVERRIDE: None
cuDNN version: 8902
Current device: 0
Library compiled for CUDA architectures: ['sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'sm_90']
GPU 0 Name: NVIDIA GeForce RTX 4090 Laptop GPU
GPU 0 Is integrated: False
GPU 0 Is multi GPU board: False
GPU 0 Multi processor count: 76
GPU 0 Total memory (GB): 15.7
GPU 0 CUDA capability (maj.min): 8.9
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by SomeUserName1**
Fixes #7567 .

### Description
 MedicalNetPerceptualSimilarity: Add multi-channel support for 3Dvolumes. 
The current version of the code in the dev branch already largely supports that besides the following:
medicalnet_* require inputs to have a single channel. 
This PR passes the multi-channel volume channel-wise to the networks and concatenates the resulting feature vectors. 
The existing code takes care of averaging over channels and spatially.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by KumoLiu** [comment]
The implementation appears to be done correctly and it successfully introduces multi-channel input support for PerceptualSimilarity. It effectively allows each channel to be processed independently, which is a substantial addition.
However, I'd like to suggest the consideration of a specific use-case scenario. The new implementation seems to be most useful under situations where independent channel processing is beneficial. Conversely, in the final computation stage, results.sum(dim=1, keepdim=True) consolidates the multi-channel perception into a single scalar evaluation. This approach works well under single modality situations.
For cases involving multi-modal inputs, it might be worth considering the separate evaluation of each modality. As a suggestion, perhaps we could add an optional flag that switches between combined evaluation and individual modality evaluation. This could potentially enhance the generalization capabilities of the implementation and provide users the flexibility to choose based on their specific use case dynamics.

**Comment by SomeUserName1** [comment]
Will adjust accordingly on monday, such that the return types are scalar or 1D array/tensor respectively.

**Comment by SomeUserName1** [comment]
For the case of LPIPS I'm not entirely sure what to do. 
There are two relevant flags, I think: `spatial` and `lpips`

Here are the relevant lines from the forward function of LPIPS:
`        if(self.lpips):
            if(self.spatial):
                res = [upsample(self.lins[kk](diffs[kk]), out_HW=in0.shape[2:]) for kk in range(self.L)]
            else:
                res = [spatial_average(self.lins[kk](diffs[kk]), keepdim=True) for kk in range(self.L)]
        else:
            if(self.spatial):
                res = [upsample(diffs[kk].sum(dim=1,keepdim=True), out_HW=in0.shape[2:]) for kk in range(self.L)]
            else:
                res = [spatial_average(diffs[kk].sum(dim=1,keepdim=True), keepdim=True) for kk in range(self.L)]`

`diffs` is the squared difference of the features.

`spatial` seems to control if spatial averaging is done in dims 2 and 3 so this is not the one we are looking for.
`lpips` controls if summing over dim 1 (i.e. the channels) is done. However the docstring states:
`lpips : bool
            [True] use linear layers on top of base/trunk network
            [False] means no linear layers; each layer is averaged together`

`lpips` false seems to do the averaging without the top layers. True does no averaging but with the input layers. 
From our perspective i think averaging should be independent on other layers, that would mean (a) another PR on lpips would be necessary to support channel-wise averaging independent of the top layers  or (b) an extra elif in the forward function for LPIPS, which does the forward pass channelwise and then either sums or returns the vector. 
@KumoLiu If i understand everything correctly, option (b) would not require chanegs upstream in lpips while (a) would keep the code in MONAI a but cleaner.
Wdyt?

**Comment by SomeUserName1** [comment]
@KumoLiu only the blossom-ci check hasn't passed. Can we merge?

File: monai/losses/perceptual.py [hunk 1]
```diff
@@ -170,9 +176,9 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
             loss = loss_sagittal + loss_axial + loss_coronal
         else:
             # 2D and real 3D cases
-            loss = self.perceptual_function(input, target)
+            loss = self.perceptual_function(input, target).squeeze()
 
-        return torch.mean(loss)
+        return torch.mean(loss, dim=0)
```
Comment (KumoLiu) [inline monai/losses/perceptual.py:181]: May I know the reason for these change? The squeeze here is for? And the loss will alway keep the channel dim?

File: monai/losses/perceptual.py [hunk 2]
```diff
@@ -74,6 +74,7 @@ def __init__(
         pretrained: bool = True,
         pretrained_path: str | None = None,
         pretrained_state_dict_key: str | None = None,
+        channelwise: bool = False,
```
[the remark this answers is not in the document]
Comment (KumoLiu) [reply monai/losses/perceptual.py:77]: Please also update the docstring, thanks.

**Review by KumoLiu — COMMENTED** [review]
Thanks for the quick update, few comments inline.

File: tests/test_perceptual_loss.py [hunk 3]
```diff
@@ -40,11 +40,26 @@
         (2, 1, 64, 64, 64),
         (2, 1, 64, 64, 64),
     ],
+    [
+        {"spatial_dims": 3, "network_type": "medicalnet_resnet10_23datasets", "is_fake_3d": False},
+        (2, 6, 64, 64, 64),
+        (2, 6, 64, 64, 64),
+    ],
+    [
+        {"spatial_dims": 3, "network_type": "medicalnet_resnet10_23datasets", "is_fake_3d": False, "channelwise": True},
```
Comment (KumoLiu) [inline tests/test_perceptual_loss.py:49]: Could you please review all the instances where "channelwise" is used? It appears there may be a misnaming issue, even with the wrong name, it doesn't affect the passing of the CI tests.

File: monai/losses/perceptual.py (hunk 1 above)
Comment (KumoLiu) [reply monai/losses/perceptual.py:184]: Hi @SomeUserName1, could you please help explain more here?
Seems the result will always be [1*C*1]？

File: tags
Comment (KumoLiu) [inline tags:1]: Added by mistake?

**Comment by KumoLiu** [comment]
Hi @SomeUserName1, I noticed that you add me review the PR again.

In case you didn't noticed the comment here. Looks a little confuse to me here. Could you please help explain it more?
https://github.com/Project-MONAI/MONAI/pull/7568#discussion_r1553071925
Thanks.

File: monai/losses/perceptual.py (hunk 1 above)
Comment (SomeUserName1) [reply monai/losses/perceptual.py:184]: Ah sorry, I just figured, that pending means I have to submit a review in files changed...

Comment (SomeUserName1) [reply monai/losses/perceptual.py]: Done

File: monai/losses/perceptual.py [hunk 4]
```diff
@@ -172,7 +181,13 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
             # 2D and real 3D cases
             loss = self.perceptual_function(input, target)
 
-        return torch.mean(loss)
+        if self.channel_wise:
+            loss = loss.squeeze()
+            loss = torch.mean(loss, dim=0)
```
Comment (KumoLiu) [inline monai/losses/perceptual.py:186]:
```suggestion
            loss = torch.mean(loss.squeeze(), dim=0)
```

**Review by KumoLiu — APPROVED** [review]
Thanks for the quick update, overall looks good to me.
Only a small comment, will merge this ASAP.
